Common/Source/Devices/devFlarm.cpp

```diff
@@ -0,0 +1,315 @@
+/*
+   LK8000 Tactical Flight Computer -  WWW.LK8000.IT
+   Released under GNU/GPL License v.2
+   See CREDITS.TXT file for authors and copyrights
+
+   $Id$
+*/
+#include "externs.h"
+#include "devFlarm.h"
+#include "dlgTools.h"
+#include "Dialogs.h"
+#include "WindowControls.h"
+#include "resource.h"
+#include "dlgIGCProgress.h"
+#include "Util/Clamp.hpp"
+#include "OS/Sleep.h"
+#include "dlgFlarmIGCDownload.h"
+
+
+
+
+
+
+static WndForm *wf = NULL;
+
+
+
+
+PDeviceDescriptor_t CDevFlarm::m_pDevice=NULL;
+BOOL CDevFlarm::m_bCompassCalOn=FALSE;
+Mutex* CDevFlarm::m_pCritSec_DeviceData=NULL;
+double CDevFlarm::m_abs_press=0.0;
+double CDevFlarm::m_delta_press=0.0;
+
+TCHAR CDevFlarm::m_szVersion[15]={0};
+
+
+
+
+bool CDevFlarm::Register(){
+	return devRegister(GetName(), cap_baro_alt|cap_vario, &Install);
+}
+
+BOOL CDevFlarm::Install( PDeviceDescriptor_t d ) {
+  StartupStore(_T("Flarm Drvier Install %s"), NEWLINE);
+	_tcscpy(d->Name, GetName());
+	d->ParseNMEA = NULL ; // ParseNMEA;
+	d->PutMacCready = NULL;
+	d->PutBugs = NULL;
+	d->PutBallast = NULL;
+	d->Open = Open;
+	d->Close = Close;
+	d->Init = NULL;
+	d->LinkTimeout = NULL;
+	d->Declare = NULL;
+	d->IsGPSSource = GetFalse;
+	d->IsBaroSource = GetTrue;
+	d->Config = Config;
+	d->ParseStream  = FlarmParseString;
+
+	return(TRUE);
+}
+
+
+#define REC_BUFFERSIZE 512
+uint8_t RingBuff[REC_BUFFERSIZE+1];
+volatile  uint16_t InCnt=0;
+volatile  uint16_t OutCnt=0;
+static bool recEnable = true;
+
+BOOL CDevFlarm::FlarmParseString(DeviceDescriptor_t *d, char *String, int len, NMEA_INFO *GPS_INFO)
+{
+if(d == NULL) return 0;
+if(String == NULL) return 0;
+if(len == 0) return 0;
+
+int cnt=0;
+
+if(recEnable)
+  while (cnt < len)
+  {
+    RingBuff[InCnt++] = (TCHAR) String[cnt++];
+    InCnt %= REC_BUFFERSIZE;
+  } //  (cnt < len)
+
+return  true;
+}
+
+
+uint8_t RecChar( DeviceDescriptor_t *d, uint8_t *inchar, uint16_t Timeout)
+{
+
+  uint16_t TimeCnt =0;
+  uint8_t Tmp;
+  while(OutCnt == InCnt)
+  {
+    Poco::Thread::sleep(1);
+    Poco::Thread::Thread::yield();
+    if(TimeCnt++ > Timeout)
+    {
+      {StartupStore(TEXT("REC_TIMEOUT_ERROR" ));}
+      return REC_TIMEOUT_ERROR;
+    }
+  }
+  Tmp = RingBuff[OutCnt++];
+  OutCnt %= REC_BUFFERSIZE;
+  if(inchar)
+    *inchar = Tmp;
+
+  return REC_NO_ERROR;
+}
+
+
+
+BOOL CDevFlarm::Open( PDeviceDescriptor_t d) {
+	m_pDevice = d;
+
+	m_pCritSec_DeviceData = new Mutex();
+
+	return TRUE;
+}
+
+BOOL CDevFlarm::Close (PDeviceDescriptor_t d) {
+	m_pDevice = NULL;
+
+	delete m_pCritSec_DeviceData;
+	m_pCritSec_DeviceData = NULL;
+
+	return TRUE;
+}
+
```

Comment:
m_pCritSec_DeviceData are useless because it's not used for protect any shared ressource
 -> Open and Close are also useless because after Mutex removing this 2 function are empty